mysql_tests/test_transaction.py

```diff
@@ -0,0 +1,265 @@
+    assert init_size == qsize(bind)
+
+    async with bind.transaction() as tx:
+        async with bind.transaction():
+            async with bind.transaction():
+                await u.update(nickname="nonono").apply()
+                assert await get_name() == "nonono"
+                tx.raise_rollback()
+                assert False, "Should not reach here"
+            assert False, "Should not reach here"
+        assert False, "Should not reach here"
+
+    assert await get_name() == "ininin"
+    assert init_size == qsize(bind)
+
+    reached = 0
+
+    async with bind.transaction():
+        async with bind.transaction() as tx:
+            async with bind.transaction():
+                await u.update(nickname="nonono").apply()
+                assert await get_name() == "nonono"
+                tx.raise_rollback()
+                assert False, "Should not reach here"
+            assert False, "Should not reach here"
+        reached += 1
+        assert await get_name() == "ininin"
+
+    assert await get_name() == "ininin"
+    assert init_size == qsize(bind)
+    assert reached == 1
+
+    async with bind.transaction():
+        async with bind.transaction() as tx:
+            async with bind.transaction():
+                await u.update(nickname="nonono").apply()
+                assert await get_name() == "nonono"
+                tx.raise_commit()
+                assert False, "Should not reach here"
+            assert False, "Should not reach here"
+        reached += 1
+        assert await get_name() == "nonono"
+
+    assert await get_name() == "nonono"
+    assert init_size == qsize(bind)
+    assert reached == 2
+
+
+# noinspection PyUnreachableCode
+async def test_end_raises_in_with(engine):
+    async with engine.transaction() as tx:
+        with pytest.raises(AssertionError, match="Illegal in managed mode"):
+            await tx.commit()
+        await tx.raise_commit()
+        assert False, "Should not reach here"
+
+    async with engine.transaction() as tx:
+        with pytest.raises(AssertionError, match="Illegal in managed mode"):
+            await tx.rollback()
+        await tx.raise_rollback()
+        assert False, "Should not reach here"
+
+
+async def test_base_exception(engine):
+    async with engine.transaction() as tx:
+        # noinspection PyBroadException
+        try:
+            await tx.raise_commit()
+        except Exception:
+            assert False, "Should not reach here"
+        assert False, "Should not reach here"
+
+
+async def test_no_rollback_on_commit_fail(engine, mocker):
+    mocker.patch("aiomysql.connection.Connection.commit").side_effect = IndexError
+    async with engine.acquire() as conn:
+        tx = await conn.transaction().__aenter__()
+        rollback = mocker.patch.object(tx._tx, "rollback")
+        with pytest.raises(IndexError):
+            await tx.__aexit__(None, None, None)
+        assert not rollback.called
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/gh/python-gino/gino/pullRequest?prid=5587950)